x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -893,33 +973,91 @@ public void findActiveTokensForRealm(String realmName, ActionListener<Collection
                 .setSize(1000)
                 .setFetchSource(true)
                 .request();
+            securityIndex.checkIndexVersionThenExecute(listener::onFailure,
+                () -> ScrollHelper.fetchAllByEntity(client, request, listener, (SearchHit hit) -> filterAndParseHit(hit, filter)));
+        }
+    }
 
+    public void findActiveTokensForUser(String username, ActionListener<Collection<Tuple<UserToken, String>>> listener) {
+        ensureEnabled();
+
+        final SecurityIndexManager frozenSecurityIndex = securityIndex.freeze();
+        if (Strings.isNullOrEmpty(username)) {
+            listener.onFailure(new IllegalArgumentException("username is required"));
+        } else if (frozenSecurityIndex.indexExists() == false) {
+            listener.onResponse(Collections.emptyList());
+        } else if (frozenSecurityIndex.isAvailable() == false) {
+            listener.onFailure(frozenSecurityIndex.getUnavailableReason());
+        } else {
+            final Instant now = clock.instant();
+            final BoolQueryBuilder boolQuery = QueryBuilders.boolQuery()
+                .filter(QueryBuilders.termQuery("doc_type", "token"))
+                .filter(QueryBuilders.boolQuery()
+                    .should(QueryBuilders.boolQuery()
+                        .must(QueryBuilders.termQuery("access_token.invalidated", false))
+                        .must(QueryBuilders.rangeQuery("access_token.user_token.expiration_time").gte(now.toEpochMilli()))
+                    )
+                    .should(QueryBuilders.termQuery("refresh_token.invalidated", false))
```

Comment:
shouldn't we add the expiration time clause for the refresh token too?

Reply:
You are right, the way the query works now, it will also return expired refresh tokens that have not yet been deleted which although not problematic from security perspective, it's unnecessary to explicitly invalidate already expired tokens. I 'll address this.